File: cpp/day_7_Coin Change 2.cpp
class Solution {
public:
    int find(vector<vector<int>> &memo, int amount, int m, vector<int> &coins){
        if(amount == 0){
            return 1;
        }
        if(amount < 0 || (m < 0 && amount >= 1)){
            return 0;
        }
        if(memo[m][amount] != -1){
            return memo[m][amount];
        }
        // cout << amount << " " << m << endl;
        memo[m][amount] = find(memo, amount, m-1, coins) + find(memo, amount-coins[m], m, coins);
        return memo[m][amount];
    }
    
    int change(int amount, vector<int>& coins) {
        vector<vector<int>> memo(coins.size(), vector<int>(amount+1, -1));
        return find(memo, amount, coins.size()-1, coins);
    }
};

File: python3/june/day_7_Coin Change 2.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        memo = [[ -1 for _ in range(amount+1)] for __ in range(len(coins))]
        
        def find(amount, m):
            
            if(amount == 0):
                return 1
            if amount < 0 or (m < 0 and amount >= 1):
                return 0
            if memo[m][amount] != -1:
                return memo[m][amount]
            # print(amount, m)
            memo[m][amount] = find(amount, m-1) + find(amount-coins[m], m)
            
            return memo[m][amount]

        return find(amount, len(coins) - 1)
